File: src/wattrouter/values.processor.ts
import { Inject, Injectable, Logger } from '@nestjs/common';
import { AsyncLocalStorage } from 'async_hooks';
import { Define, Queue } from 'agenda-nest';
import { ConfigService } from '@nestjs/config';
import { Reading } from '../enmon/upload-reading.schema.js';
import { Host } from '../common/host.js';
import { WATTroutersDiscovery } from './discovery.service.js';
import { WATTrouterModel } from './model.js';
import { ConfigWattRouter } from './config.schema.js';
import { UploadReadingRepository } from '../enmon/upload-reading.repository.js';
import { FETCH_JOB_NAME, WATTROUTER_QUEUE_NAME } from './constants.js';
import { Config } from '../config/schemas.js';
import { RegistersMapper } from './registers-mapper.js';

@Injectable()
@Queue(WATTROUTER_QUEUE_NAME)
export class WATTrouterValuesProcessor {
  private readonly logger = new Logger(WATTrouterValuesProcessor.name);

  constructor(
    @Inject(ConfigService)
    private readonly config: ConfigService<Config, true>,
    @Inject(AsyncLocalStorage)
    private readonly als: AsyncLocalStorage<Host<{ configIndex?: number; targetIndex?: number }>>,
    @Inject(WATTroutersDiscovery)
    private readonly adapters: WATTroutersDiscovery,
    @Inject(UploadReadingRepository)
    private readonly uploadJobQueue: UploadReadingRepository,
  ) {}

  @Define(FETCH_JOB_NAME)
  async handleFetchJob() {
    this.logger.log('processing fetch job...');

    const configs = this.config.getOrThrow('wattrouters', { infer: true });

    if (configs.length === 0) {
      this.logger.warn('no WATTrouters configured!');
      return;
    }

    await this.als.run(new Host({}), async () => {
      await Promise.all(
        configs.map((config, index) =>
          this.als.run(new Host({ ...this.als.getStore(), configIndex: index }), () =>
            this.processConfig(config).catch(reason => this.handleConfigProcessingError(reason)),
          ),
        ),
      );
    });

    this.logger.log(`job completed`);
  }

  private async processConfig(config: ConfigWattRouter) {
    this.logger.log(`processing config...`);

    const defaultModel = WATTrouterModel.Mx;

    if (!config.model) {
      this.logger.warn(`model not set in config, falling back to ${defaultModel}`);
    }

    const adapter = this.adapters.getByModel(config.model ?? defaultModel);

    if (!adapter) {
      this.logger.error(`adapter not found by model ${config.model}`);
      return;
    }

    this.logger.log(`fetching values from ${config.baseURL} ...`);

    const { readings, unusedMappings } = RegistersMapper.remap(
      config.mapping,
      await adapter.getReadings(config.baseURL),
    );

    if (unusedMappings.length) {
      this.logger.warn(`unused registers mapping`, { unusedMappings });
    }

    await Promise.all(
      config.targets.map((target, targetIndex) =>
        this.als.run(new Host({ ...this.als.getStore(), targetIndex }), () =>
          this.processTarget(target, readings).catch(reason => this.handleTargetProcessingError(reason)),
        ),
      ),
    );

    this.logger.log('config processed');
  }

  private async processTarget(target: ConfigWattRouter['targets'][number], readings: readonly Reading[]) {
    this.logger.log(`mapping ${readings.length} readings to jobs...`);

    const jobs = readings.map(reading => ({
      reading,
      config: target,
    }));

    this.logger.log(`pushing ${jobs.length} jobs to queue...`);

    await this.uploadJobQueue.addBulk(jobs);

    this.logger.log('jobs pushed to queue');

    await this.uploadJobQueue.scheduleInstantProcessing();

    this.logger.log('target processed');
  }

  private handleTargetProcessingError(error: unknown) {
    this.logger.error(error);
  }

  private handleConfigProcessingError(error: unknown) {
    this.logger.error(error);
  }
}

File: dotnet/App/WATTrouter/ValuesProcessor.cs
using HekyLab.EnmonAdapter.Enmon;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace HekyLab.EnmonAdapter.WATTrouter;

public class ValuesProcessor(
  ILogger<ValuesProcessor> logger,
  IAdapterSelector adapterSelector,
  IOptions<IReadOnlyCollection<Config>> _configs,
  IMeasurementsQueue measurementQueue
  )
{
  IReadOnlyCollection<Config> Configs => _configs.Value ?? [];

  public async Task HandleFetchJob(CancellationToken cancellationToken)
  {
    logger.LogInformation("handling fetch job...");

    if (Configs.Count == 0)
    {
      logger.LogWarning("no WATTrouters configured!");
      return;
    }

    await Task.WhenAll(
      Configs.Select(async (config, index) =>
      {
        using (logger.BeginScope(new { ConfigIndex = index }))
        {
          try
          {
            await ProcessConfig(config, cancellationToken);
          }
          catch (Exception e)
          {
            logger.LogError(e, "failed to process config");
          }
        }
      })
    );

    logger.LogInformation("scheduling immediate measurements processing...");

    measurementQueue.ScheduleInstantProcessing();

    logger.LogInformation("scheduler, job handling completed");
  }

  async Task ProcessConfig(Config config, CancellationToken cancellationToken)
  {
    logger.LogInformation("processing config...");

    string defaultModel = "Mx";

    if (config.Model is null)
      logger.LogWarning("model not set in config, falling back to {defaultModel}", defaultModel);

    var selectedModel = config.Model ?? defaultModel;

    IAdapter? adapter = adapterSelector.GetAdapter(selectedModel);

    if (adapter is null)
    {
      logger.LogError("adapter not found by model {Model}", selectedModel);
      return;
    }

    logger.LogInformation("fetching values from {Source}", config.Source);

    var readings = await adapter.GetReadings(config.Source, cancellationToken);

    await Task.WhenAll(config.Targets.Select(async (target, targetIndex) =>
    {
      using (logger.BeginScope(new { TargetIndex = targetIndex }))
      {
        try
        {
          await ProcessTarget(target, readings, cancellationToken);
        }
        catch (Exception e)
        {
          logger.LogError(e, "failed to process target");
        }
      }
    }));

    logger.LogInformation("config processed");
  }

  async Task ProcessTarget(Enmon.Config target, IReadOnlyCollection<Measurement> readings, CancellationToken cancellationToken)
  {
    logger.LogInformation("mapping {Count} readings to jobs...", readings.Count);

    var jobs = readings.Select(reading => new MeasurementUploadData
    {
      Reading = reading,
      Config = target,
    });

    logger.LogInformation("pushing {Count} jobs to queue...", jobs.Count());

    await measurementQueue.Push(jobs, cancellationToken);

    logger.LogInformation("jobs pushed to queue, target processed");
  }
}
